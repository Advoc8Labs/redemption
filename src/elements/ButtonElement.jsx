import React, { useState } from 'react';

import Button from 'react-bootstrap/Button';
import Delta from 'quill-delta';
import Spinner from 'react-bootstrap/Spinner';
import generateNodes from '../components/TextNodes';

function ButtonElement({
    element,
    values,
    handleRedirect,
    submit,
    onRepeatClick,
    setSubmitRef
}) {
    const [showSpinner, setShowSpinner] = useState(false);

    const elementID = element.id;
    const delta = new Delta(element.text_formatted);
    const nodes = generateNodes({
        element,
        delta,
        values,
        handleRedirect,
        elementID
    });

    async function buttonOnClick() {
        if (element.link === 'none') {
            return;
        }

        if (
            ['add_repeated_row', 'remove_repeated_row'].includes(element.link)
        ) {
            onRepeatClick();
            return;
        }

        if (element.link === 'submit') {
            setShowSpinner(true);
        }

        // Perform the submit callback
        // Note: We only need to set the spinner if the submit request failed
        // If the request succeeded then the button unmounts and calling setShowSpinner is an error
        try {
            const metadata = {
                elementType: 'button',
                elementIDs: [elementID],
                trigger: 'click'
            };
            let newStep;
            if (element.link === 'submit') {
                newStep = await submit(metadata, element.repeat || 0);
            } else {
                newStep = handleRedirect({ metadata });
            }

            // If the submit failed we want to throw here to turn off the spinner
            if (!newStep) setShowSpinner(false);
        } catch {
            setShowSpinner(false);
        }
    }
    if (element.link === 'submit') setSubmitRef(buttonOnClick);

    const styles = element.applyStyles;
    return (
        <div
            css={{
                display: 'flex',
                flexDirection: 'column',
                ...styles.getLayout(),
                ...styles.getTarget('container')
            }}
        >
            <Button
                id={element.id}
                key={element.id}
                style={{
                    display: 'flex',
                    justifyContent: 'center',
                    alignItems: 'center',
                    cursor: element.link === 'none' ? 'default' : 'pointer',
                    boxShadow: 'none',
                    maxWidth: '100%'
                }}
                css={{
                    '&:disabled': {
                        cursor: 'default !important'
                    },
                    '&:active': styles.getTarget('buttonActive'),
                    '&:hover:enabled': styles.getTarget('buttonHover'),
                    ...styles.getTarget('button')
                }}
                disabled={element.link === 'none'}
                onClick={buttonOnClick}
            >
                <div style={{ display: 'flex', position: 'relative' }}>
                    {nodes}
                    {element.image_url && (
                        <img
                            src={element.image_url}
                            style={{
                                objectFit: 'contain',
                                width: '100%',
                                height: '100%'
                            }}
                        />
                    )}
                    {showSpinner && (
                        <Spinner
                            animation='border'
                            style={{
                                color: 'white',
                                position: 'absolute',
                                top: '50%',
                                bottom: '50%',
                                marginTop: 'auto',
                                marginBottom: 'auto',
                                border: '0.2em solid currentColor',
                                borderRightColor: 'transparent'
                            }}
                            css={{
                                ...styles.getTarget('spinner'),
                                '@-webkit-keyframes spinner-border': {
                                    to: {
                                        WebkitTransform: 'rotate(360deg)',
                                        transform: 'rotate(360deg)'
                                    }
                                },
                                '@keyframes spinner-border': {
                                    to: {
                                        WebkitTransform: 'rotate(360deg)',
                                        transform: 'rotate(360deg)'
                                    }
                                },
                                '&.spinner-border': {
                                    borderRadius: '50%',
                                    animation:
                                        '0.75s linear infinite spinner-border'
                                }
                            }}
                        />
                    )}
                </div>
            </Button>
        </div>
    );
}

export default ButtonElement;
