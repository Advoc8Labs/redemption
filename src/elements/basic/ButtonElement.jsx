import React, { useMemo } from 'react';

import ReactButton from 'react-bootstrap/Button';
import TextNodes from '../components/TextNodes';
import Spinner from 'react-bootstrap/Spinner';

function adjustColor(color, amount) {
  return (
    '#' +
    color
      .replace(/^#/, '')
      .replace(/../g, (color) =>
        (
          '0' +
          Math.min(255, Math.max(0, parseInt(color, 16) + amount)).toString(16)
        ).substr(-2)
      )
  );
}

function applyButtonStyles(element, applyStyles) {
  applyStyles.addTargets('button', 'buttonActive', 'buttonHover', 'loader');

  applyStyles.apply('button', 'background_color', (a) => ({
    backgroundColor: `#${a}`
  }));
  applyStyles.applyHeight('button');
  applyStyles.applyWidth('button');
  applyStyles.applyCorners('button');
  applyStyles.applyBorders('button');
  applyStyles.applyMargin('button');

  applyStyles.applyBorders('buttonHover', 'hover_');
  if (element.properties.link !== 'none') {
    applyStyles.apply('buttonHover', 'background_color', (a) => {
      const color = `${adjustColor(a, -30)} !important`;
      return {
        backgroundColor: color,
        borderColor: color,
        transition: 'background 0.3s !important'
      };
    });
  }
  if (element.styles.hover_background_color) {
    applyStyles.apply('buttonHover', 'hover_background_color', (a) => ({
      backgroundColor: `#${a} !important`
    }));
  }

  applyStyles.applyBorders('buttonActive', 'selected_');
  if (element.styles.selected_background_color) {
    applyStyles.apply('buttonHover', 'selected_background_color', (a) => ({
      backgroundColor: `#${a} !important`
    }));
  }

  applyStyles.apply('loader', ['height', 'height_unit'], (a, b) => {
    const thirdHeight = Math.round(a / 3);
    const dimension = `${thirdHeight}${b}`;
    return { width: dimension, height: dimension };
  });

  return applyStyles;
}

function ButtonElement({
  element,
  applyStyles,
  values = null,
  loader = null,
  handleRedirect = () => {},
  onClick = () => {},
  elementProps = {},
  children
}) {
  const styles = useMemo(() => applyButtonStyles(element, applyStyles), [
    applyStyles
  ]);

  return (
    <ReactButton
      id={element.id}
      key={element.id}
      style={{
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        cursor: element.properties.link === 'none' ? 'default' : 'pointer',
        boxShadow: 'none',
        maxWidth: '100%',
        position: 'relative'
      }}
      css={{
        '&:disabled': { cursor: 'default !important' },
        '&:active': styles.getTarget('buttonActive'),
        '&:hover:enabled': styles.getTarget('buttonHover'),
        '&&': styles.getTarget('button')
      }}
      disabled={element.properties.link === 'none' || loader}
      onClick={onClick}
      {...elementProps}
    >
      {loader ? (
        <div css={styles.getTarget('loader')}>{loader}</div>
      ) : (
        <>
          {element.properties.image && (
            <img
              src={element.properties.image}
              style={{
                objectFit: 'contain',
                maxWidth: '80%',
                maxHeight: '100%'
              }}
            />
          )}
          <TextNodes
            element={element}
            values={values}
            applyStyles={applyStyles}
            handleRedirect={handleRedirect}
          />
        </>
      )}
      {children}
    </ReactButton>
  );
}

export default ButtonElement;
export { adjustColor };
