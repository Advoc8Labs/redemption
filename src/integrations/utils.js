import { installPlaid } from './plaid';
import { emailLogin, installFirebase } from './firebase';
import { initializeTagManager } from './googleTagManager';

export function dynamicImport(dependency, async = true, index = 0) {
  if (async) {
    return new Promise((resolve) => {
      global.scriptjsLoadPromise.then(($script) => {
        $script.default(dependency, resolve);
      });
    });
  } else if (index < dependency.length) {
    return new Promise((resolve) => {
      global.scriptjsLoadPromise.then(($script) => {
        $script.default(dependency[index], resolve);
      });
    }).then(() => dynamicImport(dependency, false, index + 1));
  }
}

export async function initializeIntegrations(integrations, clientArg, init) {
  const gtm = integrations['google-tag-manager'];
  const fb = integrations.firebase;
  const plaid = integrations.plaid;

  const [, firebase] = await Promise.all([
    installPlaid(!!plaid),
    installFirebase(fb)
  ]);

  if (gtm) initializeTagManager(gtm);
  if (fb && !init) emailLogin(fb, firebase, clientArg);
}
