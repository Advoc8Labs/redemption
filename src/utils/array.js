/**
 * Inserts an element into a list without side effects.
 */
function justInsert(list, element, index, replace = true) {
    return [
        ...list.slice(0, index),
        element,
        ...list.slice(replace ? index + 1 : index)
    ];
}

/**
 * Removes an element from a list without side effects.
 */
function justRemove(list, index) {
    return [...list.slice(0, index), ...list.slice(index + 1)];
}

function toList(itemOrList) {
    return Array.isArray(itemOrList) ? itemOrList : [itemOrList];
}

export { justInsert, justRemove, toList };
