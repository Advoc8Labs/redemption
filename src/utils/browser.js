export function runningInClient() {
  return typeof window === 'object';
}
