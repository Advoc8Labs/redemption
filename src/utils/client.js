import * as errors from './error';
import { initInfo } from './init';
import encodeGetParams from './string';

export default class Client {
    async begin(formKey) {
        const { userKey, apiKey } = initInfo();
        const params = encodeGetParams({
            flow_key: formKey,
            ...(userKey ? { fuser_key: userKey } : {})
        });
        const url = `https://api.feathery.tech/api/panel/step/?${params}`;
        const options = {
            cache: 'no-store',
            headers: { Authorization: 'Token ' + apiKey }
        };
        return fetch(url, options).then((response) => {
            const { status } = response;
            switch (status) {
                case 200:
                    return response.json();
                case 401:
                    return Promise.reject(
                        new errors.APIKeyError('Invalid API key')
                    );
                case 404:
                    return Promise.reject(
                        new errors.UserKeyError('Invalid User key')
                    );
                default:
                    return Promise.reject(
                        new errors.FetchError('Unknown error')
                    );
            }
        });
    }

    async submitStep(formKey, stepNum, servars, action) {
        // servars = [{key: <servarKey>, <type>: <value>}]
        const { userKey, apiKey } = initInfo();
        const url = `https://api.feathery.tech/api/panel/step/submit/`;
        const data = {
            ...(userKey ? { fuser_key: userKey } : {}),
            flow_key: formKey,
            step_number: stepNum,
            servars: action === 'next' ? servars : [],
            action
        };
        const options = {
            cache: 'no-store',
            headers: {
                Authorization: 'Token ' + apiKey,
                'Content-Type': 'application/json'
            },
            method: 'POST',
            body: JSON.stringify(data)
        };
        return fetch(url, options).then((response) => {
            const { status } = response;
            switch (status) {
                case 200:
                    return response.json();
                case 201:
                    return response.json();
                case 401:
                    return Promise.reject(
                        new errors.APIKeyError('Invalid API key')
                    );
                case 404:
                    return Promise.reject(
                        new errors.UserKeyError('Invalid User key')
                    );
                default:
                    return Promise.reject(
                        new errors.FetchError('Unknown error')
                    );
            }
        });
    }
}
