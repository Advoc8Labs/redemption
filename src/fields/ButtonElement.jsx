import React, { useState } from 'react';
import Button from 'react-bootstrap/Button';
import Spinner from 'react-bootstrap/Spinner';
import Delta from 'quill-delta';

import { adjustColor, alignmentMap } from '../utils/formHelperFunctions';
import generateNodes from './textNodes';
import { borderStyleFromField, marginStyleFromField } from '../utils/styles';

function ButtonElement({
    field,
    fieldValues,
    addRepeatedRow,
    removeRepeatedRow,
    submit,
    setSubmitRef
}) {
    const [showSpinner, setShowSpinner] = useState(false);

    const elementID = field.id;
    const repeat = field.repeat || 0;
    const delta = new Delta(field.text_formatted);
    const nodes = generateNodes({
        delta,
        fieldValues,
        field,
        submit,
        repeat,
        elementID
    });

    async function buttonOnClick() {
        if (field.link === 'none') {
            return;
        }

        if (field.link === 'add_repeated_row') {
            addRepeatedRow();
            return;
        }

        if (field.link === 'remove_repeated_row') {
            removeRepeatedRow(field.repeat);
            return;
        }

        if (field.link === 'submit' && field.show_spinner_on_submit) {
            setShowSpinner(true);
        }

        // Perform the submit callback
        // Note: We only need to set the spinner if the submit request failed
        // If the request succeeded then the button unmounts and calling setShowSpinner is an error
        try {
            const newStep = await submit(
                field.link === 'submit',
                {
                    elementType: 'button',
                    elementIDs: [elementID],
                    trigger: 'click'
                },
                repeat
            );

            // If the submit failed we want to throw here to turn off the spinner
            if (!newStep) setShowSpinner(false);
        } catch {
            setShowSpinner(false);
        }
    }
    if (field.link === 'submit') setSubmitRef(buttonOnClick);

    let hoverStyles = borderStyleFromField(field, 'hover_');
    if (field.link !== 'none') {
        hoverStyles = {
            backgroundColor: `${adjustColor(
                field.button_color,
                -30
            )} !important`,
            borderColor: `${adjustColor(field.button_color, -30)} !important`,
            transition: 'background 0.3s !important',
            ...hoverStyles
        };
    }
    if (field.hover_background_color)
        hoverStyles.backgroundColor = `#${field.hover_background_color} !important`;
    const selectedStyles = borderStyleFromField(field, 'selected_');
    if (field.selected_background_color)
        selectedStyles.backgroundColor = `#${field.selected_background_color} !important`;

    const borderRadius = `${field.corner_top_left_radius}px ${field.corner_top_right_radius}px ${field.corner_bottom_right_radius}px ${field.corner_bottom_left_radius}px`;
    const halfHeight = Math.round(field.button_height / 2);
    return (
        <div
            css={{
                gridColumnStart: field.column_index + 1,
                gridRowStart: field.row_index + 1,
                gridColumnEnd: field.column_index_end + 2,
                gridRowEnd: field.row_index_end + 2,
                display: 'flex',
                flexDirection: 'column',
                alignItems: alignmentMap[field.layout],
                textAlign: field.layout,
                justifyContent: field.vertical_layout
            }}
        >
            <Button
                id={field.id}
                key={field.id}
                style={{
                    display: 'flex',
                    justifyContent: 'center',
                    alignItems: 'center',
                    cursor: field.link === 'none' ? 'default' : 'pointer',
                    backgroundColor: `#${field.button_color}`,
                    boxShadow: 'none',
                    height: `${field.button_height}${field.button_height_unit}`,
                    width: `${field.button_width}${field.button_width_unit}`,
                    maxWidth: '100%',
                    borderRadius,
                    ...borderStyleFromField(field),
                    ...marginStyleFromField(field)
                }}
                css={{
                    '&:active': selectedStyles,
                    '&:hover:enabled': hoverStyles,
                    '&:disabled': {
                        cursor: 'default !important'
                    }
                }}
                disabled={field.link === 'none'}
                onClick={buttonOnClick}
            >
                <div style={{ display: 'flex', position: 'relative' }}>
                    {nodes}
                    {field.image_url && (
                        <img
                            src={field.image_url}
                            style={{
                                objectFit: 'contain',
                                width: '100%',
                                height: '100%'
                            }}
                        />
                    )}
                    {showSpinner && (
                        <Spinner
                            animation='border'
                            style={{
                                color: 'white',
                                position: 'absolute',
                                right: `-${field.button_height}${field.button_height_unit}`,
                                width: `${halfHeight}${field.button_height_unit}`,
                                height: `${halfHeight}${field.button_height_unit}`,
                                border: '0.2em solid currentColor',
                                borderRightColor: 'transparent'
                            }}
                        />
                    )}
                </div>
            </Button>
        </div>
    );
}

export default ButtonElement;
