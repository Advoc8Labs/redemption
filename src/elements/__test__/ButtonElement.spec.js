import React from 'react';
import { create, act } from 'react-test-renderer';
import ButtonElement from '../ButtonElement';
import { getButtonStyles } from '../../utils/styles';

describe('ButtonElement', () => {
    it('renders an empty Button element', async () => {
        // Arrange
        const props = {
            element: {
                text_formatted: [{ insert: '\n' }],
                column_index: 0,
                row_index: 0,
                column_index_end: 0,
                row_index_end: 0,
                styles: {
                    border_top_color: '2954af',
                    border_right_color: '2954af',
                    border_bottom_color: '2954af',
                    border_left_color: '2954af',
                    background_color: '2954af',
                    layout: 'center'
                },
                mobile_styles: {}
            },
            values: {},
            submit: () => {},
            setSubmitRef: () => {},
            onRepeatClick: () => {}
        };
        getButtonStyles(props.element);

        // Act
        let button;
        act(() => {
            button = create(<ButtonElement {...props} />);
        });
        const tree = button.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });

    it('renders a button with plain text', async () => {
        // Arrange
        const props = {
            element: {
                text_formatted: [{ insert: 'Hello World!' }],
                column_index: 0,
                row_index: 0,
                column_index_end: 0,
                row_index_end: 0,
                styles: {
                    background_color: '2954af',
                    layout: 'center'
                },
                mobile_styles: {}
            },
            values: {},
            submit: () => {},
            setSubmitRef: () => {},
            onRepeatClick: () => {}
        };
        getButtonStyles(props.element);

        // Act
        let button;
        act(() => {
            button = create(<ButtonElement {...props} />);
        });
        const tree = button.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });

    it('renders a button with custom styles', async () => {
        // Arrange
        const props = {
            element: {
                text_formatted: [
                    {
                        insert: 'Hello World!',
                        attributes: {
                            weight: 700,
                            color: '#000000FF',
                            size: '48px',
                            font: 'sans-serif',
                            italic: true,
                            strike: false,
                            underline: false
                        }
                    }
                ],
                column_index: 0,
                row_index: 0,
                column_index_end: 0,
                row_index_end: 0,
                styles: {
                    background_color: '2954af',
                    layout: 'center'
                },
                mobile_styles: {}
            },
            values: {},
            submit: () => {},
            setSubmitRef: () => {},
            onRepeatClick: () => {}
        };
        getButtonStyles(props.element);

        // Act
        let button;
        act(() => {
            button = create(<ButtonElement {...props} />);
        });
        const tree = button.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });
});
