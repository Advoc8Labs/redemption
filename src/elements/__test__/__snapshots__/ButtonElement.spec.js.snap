// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonElement renders a button with custom styles 1`] = `
<div
  className="css-1wlrtcl-ButtonElement"
>
  <button
    className="css-1wth4m0-ButtonElement btn btn-primary"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "boxShadow": "none",
        "cursor": "pointer",
        "display": "flex",
        "justifyContent": "center",
        "maxWidth": "100%",
      }
    }
    type="button"
  >
    <div
      style={
        Object {
          "display": "flex",
          "position": "relative",
        }
      }
    >
      <span
        className="css-1k1lbqh-generateNodes"
        onClick={[Function]}
      >
        Hello World!
      </span>
    </div>
  </button>
</div>
`;

exports[`ButtonElement renders a button with plain text 1`] = `
<div
  className="css-1wlrtcl-ButtonElement"
>
  <button
    className="css-1wth4m0-ButtonElement btn btn-primary"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "boxShadow": "none",
        "cursor": "pointer",
        "display": "flex",
        "justifyContent": "center",
        "maxWidth": "100%",
      }
    }
    type="button"
  >
    <div
      style={
        Object {
          "display": "flex",
          "position": "relative",
        }
      }
    >
      <span
        className="css-enkqi-generateNodes"
        onClick={[Function]}
      >
        Hello World!
      </span>
    </div>
  </button>
</div>
`;

exports[`ButtonElement renders an empty Button element 1`] = `
<div
  className="css-1wlrtcl-ButtonElement"
>
  <button
    className="css-jnnik-ButtonElement btn btn-primary"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "boxShadow": "none",
        "cursor": "pointer",
        "display": "flex",
        "justifyContent": "center",
        "maxWidth": "100%",
      }
    }
    type="button"
  >
    <div
      style={
        Object {
          "display": "flex",
          "position": "relative",
        }
      }
    >
      <span
        className="css-enkqi-generateNodes"
        onClick={[Function]}
      >
        

      </span>
    </div>
  </button>
</div>
`;
